pkg/agent/controller/egress/egress_controller.go

```diff
@@ -0,0 +1,674 @@
+			klog.Errorf("DeletedFinalStateUnknown contains non-Egress object: %v", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(egress.Name)
+}
+
+// Run will create defaultWorkers workers (go routines) which will process the Node events from the
+// workqueue.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.egressListerSynced) {
+		return
+	}
+
+	go wait.NonSlidingUntil(c.watch, 5*time.Second, stopCh)
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the workqueue.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect strings (Egress name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncEgress(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing Egress %s, requeuing. Error: %v", key, err)
+	}
+	return true
+}
+
+// getOrAllocateMark gets or allocates a mark for a local Egress. Multiple Egresses can share same Egress IP.
+// The mark is per Egress IP, not per Egress.
+func (c *Controller) getOrAllocateMark(egressName, egressIP string) (uint32, error) {
+	c.egressIPStatesMutex.Lock()
+	defer c.egressIPStatesMutex.Unlock()
+
+	ipState, exists := c.egressIPStates[egressIP]
+	// Create an egressIPState if this is the first Egress using the IP.
+	if !exists {
+		ipState = &egressIPState{
+			egressIP:    net.ParseIP(egressIP),
+			egressNames: sets.NewString(egressName),
+		}
+		c.egressIPStates[egressIP] = ipState
+	} else if !ipState.egressNames.Has(egressName) {
+		ipState.egressNames.Insert(egressName)
+	}
+
+	// Ensure the Egress IP has a mark allocated.
+	var err error
+	if ipState.mark == 0 {
+		ipState.mark, err = c.idAllocator.allocate()
+		if err != nil {
+			return 0, fmt.Errorf("error allocating mark for IP %s: %v", egressIP, err)
+		}
+	}
+	// Ensure datapath is configured properly.
+	if !ipState.datapathInstalled {
+		if err := c.ofClient.InstallSNATMarkFlows(ipState.egressIP, ipState.mark); err != nil {
```

Comment:
Could we install flows / rule in the caller with the lock released?

Reply:
It may introduce extra external calls. For example, if two new Egresses have same Egress IP, and the first one doesn't configure the datapath with the lock, the other one will also try to configure the datapath, unless we introduce per-IP locks. However, even if we add per-IP locks, at least iptables calls will be serial anyway. Do you think it's worthwhile?

Reply:
The per-IP lock will be necessary as there could be race condition when an Egress starts to use an EgressIP while another Egress stops to use the same EgressIP.